File: api/app/domains/crypto/reference/grants.py
"""Sender key distribution: wrapping a chain key to each recipient device.

A sender that wants to write into an epoch generates a chain key, then wraps it once per recipient
device using ephemeral X25519 ECDH against that device's public key. The server stores the wrapped
blobs and can open none of them.

Two things here carry most of the security weight:

**The distribution signature.** A grant on its own says nothing about who produced it. The sender
signs (chat, epoch, sender_key_id, chain signing key, start index) with its long-term Ed25519
identity, so a recipient can tell a real distribution from one the server fabricated.

**The member set hash.** An epoch commits to the exact set of devices it was created for. Before
wrapping, a client recomputes the hash from the roster the server returned and refuses on mismatch.
Without that check a malicious server silently adds a ghost device and every sender dutifully wraps
the chain key for it — the cryptography behaving perfectly while confidentiality is lost, because
membership is server-authoritative while confidentiality is client-enforced. The backend cannot
enforce this; it only stores the value so clients can compare.
"""
import os
from hashlib import sha256

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from app.domains.crypto.reference.primitives import (
    DS_GRANT,
    DS_MEMBER_SET,
    DS_SENDER_KEY,
    b64u_decode,
    b64u_encode,
    u32be,
    uuid_bytes,
)

WRAP_ALGORITHM = "x25519_hkdf_sha256_aes256gcm_v1"
SENDER_KEY_ALGORITHM = "hkdf_sha256_aes256gcm_v1"

NONCE_BYTES = 12
WRAP_KEY_BYTES = 32


def compute_member_set_hash(device_ids) -> str:
    """Commit to an epoch's exact device set.

    Sorted so every participant derives the same value regardless of roster ordering.
    """
    joined = "|".join(sorted(str(d) for d in device_ids)).encode("utf-8")
    return sha256(DS_MEMBER_SET + joined).hexdigest()


def distribution_signing_payload(
        chat_id, epoch: int, sender_key_id, signing_public: bytes, chain_start_index: int
) -> bytes:
    return (
        DS_SENDER_KEY
        + uuid_bytes(chat_id)
        + u32be(epoch)
        + uuid_bytes(sender_key_id)
        + signing_public
        + u32be(chain_start_index)
    )


def sign_distribution(
        *, identity_signing_private: bytes, chat_id, epoch: int, sender_key_id,
        chain_signing_public: bytes, chain_start_index: int,
) -> bytes:
    """Vouch for a chain's signing key with the sender's long-term identity key."""
    return Ed25519PrivateKey.from_private_bytes(identity_signing_private).sign(
        distribution_signing_payload(
            chat_id, epoch, sender_key_id, chain_signing_public, chain_start_index
        )
    )


def verify_distribution(
        *, identity_signing_public: bytes, signature: bytes, chat_id, epoch: int, sender_key_id,
        chain_signing_public: bytes, chain_start_index: int,
) -> bool:
    try:
        Ed25519PublicKey.from_public_bytes(identity_signing_public).verify(
            signature,
            distribution_signing_payload(
                chat_id, epoch, sender_key_id, chain_signing_public, chain_start_index
            ),
        )
        return True
    except (InvalidSignature, ValueError, TypeError):
        return False


def build_grant_aad(
        *, chat_id, epoch: int, sender_key_id, sender_device_id, recipient_device_id,
        ephemeral_public: bytes,
) -> bytes:
    """Bind a wrapped key to exactly one (chat, epoch, sender, recipient, ephemeral) tuple.

    Including the ephemeral public key in the AAD *and* the HKDF info blocks reuse of a captured
    ephemeral against a different recipient.
    """
    return (
        DS_GRANT
        + uuid_bytes(chat_id)
        + u32be(epoch)
        + uuid_bytes(sender_key_id)
        + uuid_bytes(sender_device_id)
        + uuid_bytes(recipient_device_id)
        + ephemeral_public
    )


def _derive_wrap_key(shared_secret: bytes, chat_id, epoch: int, aad: bytes) -> tuple[bytes, bytes]:
    okm = HKDF(
        algorithm=SHA256(),
        length=WRAP_KEY_BYTES + NONCE_BYTES,
        salt=sha256(uuid_bytes(chat_id) + u32be(epoch)).digest(),
        info=aad,
    ).derive(shared_secret)

    return okm[:WRAP_KEY_BYTES], okm[WRAP_KEY_BYTES:]


def wrap_chain_key(
        *, chain_key: bytes, chain_start_index: int, recipient_public: bytes,
        chat_id, epoch: int, sender_key_id, sender_device_id, recipient_device_id,
) -> tuple[str, str]:
    """Wrap a chain key for one recipient device. Returns (ephemeral_public, wrapped) as b64u.

    `recipient_public` should be the recipient's signed prekey when available — that gives forward
    secrecy for the grant once the prekey rotates — falling back to their long-term identity key.
    """
    ephemeral_private = X25519PrivateKey.generate()
    ephemeral_public = ephemeral_private.public_key().public_bytes(
        encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
    )

    shared_secret = ephemeral_private.exchange(
        X25519PublicKey.from_public_bytes(recipient_public)
    )

    aad = build_grant_aad(
        chat_id=chat_id, epoch=epoch, sender_key_id=sender_key_id,
        sender_device_id=sender_device_id, recipient_device_id=recipient_device_id,
        ephemeral_public=ephemeral_public,
    )
    wrap_key, nonce = _derive_wrap_key(shared_secret, chat_id, epoch, aad)

    ciphertext = AESGCM(wrap_key).encrypt(
        nonce, chain_key + u32be(chain_start_index), aad
    )

    return b64u_encode(ephemeral_public), b64u_encode(nonce + ciphertext)


def unwrap_chain_key(
        *, wrapped: str, ephemeral_public: str, recipient_private: bytes,
        chat_id, epoch: int, sender_key_id, sender_device_id, recipient_device_id,
) -> tuple[bytes, int]:
    """Inverse of wrap_chain_key. Returns (chain_key, chain_start_index).

    Any mismatch in the bound fields surfaces as a GCM tag failure rather than a wrong-but-usable
    key, so a relocated grant fails closed.
    """
    ephemeral_raw = b64u_decode(ephemeral_public)
    blob = b64u_decode(wrapped)
    nonce, ciphertext = blob[:NONCE_BYTES], blob[NONCE_BYTES:]

    shared_secret = X25519PrivateKey.from_private_bytes(recipient_private).exchange(
        X25519PublicKey.from_public_bytes(ephemeral_raw)
    )

    aad = build_grant_aad(
        chat_id=chat_id, epoch=epoch, sender_key_id=sender_key_id,
        sender_device_id=sender_device_id, recipient_device_id=recipient_device_id,
        ephemeral_public=ephemeral_raw,
    )
    wrap_key, _ = _derive_wrap_key(shared_secret, chat_id, epoch, aad)

    plaintext = AESGCM(wrap_key).decrypt(nonce, ciphertext, aad)

    return plaintext[:32], int.from_bytes(plaintext[32:36], "big")

File: front/src/app/core/crypto/grants.ts
import { ed25519, x25519 } from '@noble/curves/ed25519.js';
import { hkdf } from '@noble/hashes/hkdf.js';
import { sha256 } from '@noble/hashes/sha2.js';

import {
    b64uDecode,
    b64uEncode,
    concatBytes,
    DS_GRANT,
    DS_MEMBER_SET,
    DS_SENDER_KEY,
    u32be,
    utf8,
    uuidBytes,
} from './primitives';

/**
 * Sender key distribution. Client half of `reference/grants.py`.
 *
 * Two things here carry most of the security weight:
 *
 * 1. The distribution signature — a grant alone says nothing about who produced it, so the sender
 *    signs the chain's signing key with its long-term identity.
 * 2. The member set hash — an epoch commits to an exact device set. Before wrapping, a client
 *    recomputes the hash from the roster and refuses on mismatch. Without that check a malicious
 *    server silently adds a ghost device and every sender dutifully wraps for it, with the
 *    cryptography behaving perfectly. **The backend cannot enforce this. It lives here.**
 */

export const WRAP_ALGORITHM = 'x25519_hkdf_sha256_aes256gcm_v1';
export const SENDER_KEY_ALGORITHM = 'hkdf_sha256_aes256gcm_v1';

const NONCE_BYTES = 12;
const WRAP_KEY_BYTES = 32;

/** Commit to an epoch's exact device set. Sorted so every participant derives the same value. */
export function computeMemberSetHash(deviceIds: readonly string[]): string {
    const joined = [...deviceIds].map(String).sort().join('|');
    const digest = sha256(concatBytes(DS_MEMBER_SET, utf8(joined)));

    return Array.from(digest)
        .map((b) => b.toString(16).padStart(2, '0'))
        .join('');
}

export function distributionSigningPayload(
    chatId: string,
    epoch: number,
    senderKeyId: string,
    signingPublic: Uint8Array,
    chainStartIndex: number
): Uint8Array {
    return concatBytes(
        DS_SENDER_KEY,
        uuidBytes(chatId),
        u32be(epoch),
        uuidBytes(senderKeyId),
        signingPublic,
        u32be(chainStartIndex)
    );
}

export function signDistribution(params: {
    identitySigningPrivate: Uint8Array;
    chatId: string;
    epoch: number;
    senderKeyId: string;
    chainSigningPublic: Uint8Array;
    chainStartIndex: number;
}): string {
    return b64uEncode(
        ed25519.sign(
            distributionSigningPayload(
                params.chatId,
                params.epoch,
                params.senderKeyId,
                params.chainSigningPublic,
                params.chainStartIndex
            ),
            params.identitySigningPrivate
        )
    );
}

export function verifyDistribution(params: {
    identitySigningPublic: Uint8Array;
    signature: string;
    chatId: string;
    epoch: number;
    senderKeyId: string;
    chainSigningPublic: Uint8Array;
    chainStartIndex: number;
}): boolean {
    try {
        return ed25519.verify(
            b64uDecode(params.signature),
            distributionSigningPayload(
                params.chatId,
                params.epoch,
                params.senderKeyId,
                params.chainSigningPublic,
                params.chainStartIndex
            ),
            params.identitySigningPublic
        );
    } catch {
        return false;
    }
}

/** Binds a wrapped key to exactly one (chat, epoch, sender, recipient, ephemeral) tuple. */
export function buildGrantAad(params: {
    chatId: string;
    epoch: number;
    senderKeyId: string;
    senderDeviceId: string;
    recipientDeviceId: string;
    ephemeralPublic: Uint8Array;
}): Uint8Array {
    return concatBytes(
        DS_GRANT,
        uuidBytes(params.chatId),
        u32be(params.epoch),
        uuidBytes(params.senderKeyId),
        uuidBytes(params.senderDeviceId),
        uuidBytes(params.recipientDeviceId),
        params.ephemeralPublic
    );
}

function deriveWrapKey(
    sharedSecret: Uint8Array,
    chatId: string,
    epoch: number,
    aad: Uint8Array
): { wrapKey: Uint8Array; nonce: Uint8Array } {
    const salt = sha256(concatBytes(uuidBytes(chatId), u32be(epoch)));
    const okm = hkdf(sha256, sharedSecret, salt, aad, WRAP_KEY_BYTES + NONCE_BYTES);

    return { wrapKey: okm.slice(0, WRAP_KEY_BYTES), nonce: okm.slice(WRAP_KEY_BYTES) };
}

async function importAesKey(raw: Uint8Array): Promise<CryptoKey> {
    return crypto.subtle.importKey('raw', raw as BufferSource, 'AES-GCM', false, ['encrypt', 'decrypt']);
}

export interface WrappedGrant {
    ephemeralPublicKey: string;
    wrappedChainKey: string;
}

/**
 * Wrap a chain key for one recipient device.
 *
 * `recipientPublic` should be the recipient's signed prekey when available — that gives forward
 * secrecy for the grant once the prekey rotates — falling back to their long-term identity key.
 */
export async function wrapChainKey(params: {
    chainKey: Uint8Array;
    chainStartIndex: number;
    recipientPublic: Uint8Array;
    chatId: string;
    epoch: number;
    senderKeyId: string;
    senderDeviceId: string;
    recipientDeviceId: string;
}): Promise<WrappedGrant> {
    const ephemeralPrivate = x25519.utils.randomSecretKey();
    const ephemeralPublic = x25519.getPublicKey(ephemeralPrivate);

    const sharedSecret = x25519.getSharedSecret(ephemeralPrivate, params.recipientPublic);

    const aad = buildGrantAad({ ...params, ephemeralPublic });
    const { wrapKey, nonce } = deriveWrapKey(sharedSecret, params.chatId, params.epoch, aad);

    const ciphertext = new Uint8Array(
        await crypto.subtle.encrypt(
            { name: 'AES-GCM', iv: nonce as BufferSource, additionalData: aad as BufferSource },
            await importAesKey(wrapKey),
            concatBytes(params.chainKey, u32be(params.chainStartIndex)) as BufferSource
        )
    );

    return {
        ephemeralPublicKey: b64uEncode(ephemeralPublic),
        wrappedChainKey: b64uEncode(concatBytes(nonce, ciphertext)),
    };
}

/** Inverse of wrapChainKey. Any mismatch in the bound fields fails closed as a GCM tag error. */
export async function unwrapChainKey(params: {
    wrapped: string;
    ephemeralPublic: string;
    recipientPrivate: Uint8Array;
    chatId: string;
    epoch: number;
    senderKeyId: string;
    senderDeviceId: string;
    recipientDeviceId: string;
}): Promise<{ chainKey: Uint8Array; chainStartIndex: number }> {
    const ephemeralPublic = b64uDecode(params.ephemeralPublic);
    const blob = b64uDecode(params.wrapped);
    const nonce = blob.slice(0, NONCE_BYTES);
    const ciphertext = blob.slice(NONCE_BYTES);

    const sharedSecret = x25519.getSharedSecret(params.recipientPrivate, ephemeralPublic);

    const aad = buildGrantAad({ ...params, ephemeralPublic });
    const { wrapKey } = deriveWrapKey(sharedSecret, params.chatId, params.epoch, aad);

    const plaintext = new Uint8Array(
        await crypto.subtle.decrypt(
            { name: 'AES-GCM', iv: nonce as BufferSource, additionalData: aad as BufferSource },
            await importAesKey(wrapKey),
            ciphertext as BufferSource
        )
    );

    return {
        chainKey: plaintext.slice(0, 32),
        chainStartIndex: new DataView(plaintext.buffer, plaintext.byteOffset + 32, 4).getUint32(0, false),
    };
}
